File: install/scripts/gen_installer.py
# ======================================================================================
# File         : CreateInstaller.py
# Author       : Wu Jie 
# Last Change  : 05/18/2009 | 21:01:02 PM | Monday,May
# Description  : 
# ======================================================================================

#/////////////////////////////////////////////////////////////////////////////
# imports
#/////////////////////////////////////////////////////////////////////////////

import os.path
import shutil
import zipfile
import settings
import misc

#/////////////////////////////////////////////////////////////////////////////
# global variables
#/////////////////////////////////////////////////////////////////////////////

# general
dest_root_path = settings.release_path
dest_version_path = os.path.join ( settings.release_path, settings.version )
installer_exe_name = "exvim-installer"

#/////////////////////////////////////////////////////////////////////////////
# functions
#/////////////////////////////////////////////////////////////////////////////

# ------------------------------------------------------------------ 
# Desc: 
# ------------------------------------------------------------------ 

def PreCheck ():
    print "" 
    print "#########################" 
    print "Pre-Checking..."
    print "#########################" 
    print ""

    print "version = %s" % settings.version

    # check source path, if not found, return false
    print "exVim_path = %s" % os.path.abspath(settings.exVim_path)
    if os.path.isdir( os.path.abspath(settings.exVim_path) ) == False :
        print "source path not found"
        return False

    # check dest path, if not found, create one
    print "dest_root_path = %s" % dest_root_path
    if os.path.isdir(dest_root_path) == False :
        print "dest path not found, create the path."
        os.makedirs (dest_root_path)
        print "dest path %s created!" % dest_root_path

    # check dest package, if not found, show warnning message
    exvim_path = os.path.join(dest_root_path,"rawdata/exvim")
    if os.path.isdir(exvim_path) == False :
        print "exvim path not found!"
        os.makedirs (exvim_path)
        print "exvim path %s created!" % exvim_path

    # check third part packages path, if not found, show warnning message

    rawvim_path = os.path.join(dest_root_path,"rawdata/exvim/vim")
    if os.path.isdir(rawvim_path) == False :
        print "Warnning: raw-vim not found! (%s)" % rawvim_path
    else :
        print "raw-vim OK!"

    diffutils_path = os.path.join(dest_root_path,"rawdata/GnuWin32/diffutils")
    if os.path.isdir(diffutils_path) == False :
        print "Warnning: diffutils not found! (%s)" % diffutils_path
    else :
        print "diffutils OK!"

    gawk_path = os.path.join(dest_root_path,"rawdata/GnuWin32/gawk")
    if os.path.isdir(gawk_path) == False :
        print "Warnning: gawk not found! (%s)" % gawk_path
    else :
        print "gawk OK!"

    idutils_path = os.path.join(dest_root_path,"rawdata/GnuWin32/id-utils")
    if os.path.isdir(idutils_path) == False :
        print "Warnning: id-utils not found! (%s)" % idutils_path
    else :
        print "id-utils OK!"

    libiconv_path = os.path.join(dest_root_path,"rawdata/GnuWin32/libiconv")
    if os.path.isdir(libiconv_path) == False :
        print "Warnning: libiconv not found! (%s)" % libiconv_path
    else :
        print "libiconv OK!"

    libintl_path = os.path.join(dest_root_path,"rawdata/GnuWin32/libintl")
    if os.path.isdir(libintl_path) == False :
        print "Warnning: libintl not found! (%s)" % libintl_path
    else :
        print "libintl OK!"

    sed_path = os.path.join(dest_root_path,"rawdata/GnuWin32/sed")
    if os.path.isdir(sed_path) == False :
        print "Warnning: sed not found! (%s)" % sed_path
    else :
        print "sed OK!"

    regex_path = os.path.join(dest_root_path,"rawdata/GnuWin32/regex")
    if os.path.isdir(regex_path) == False :
        print "Warnning: regex not found! (%s)" % regex_path
    else :
        print "regex OK!"

    srchighlite_path = os.path.join(dest_root_path,"rawdata/GnuWin32/src-highlite")
    if os.path.isdir(srchighlite_path) == False :
        print "Warnning: src-highlite not found! (%s)" % srchighlite_path
    else :
        print "src-highlite OK!"

    graphviz_path = os.path.join(dest_root_path,"rawdata/Graphviz")
    if os.path.isdir(graphviz_path) == False :
        print "Warnning: Graphviz not found! (%s)" % graphviz_path
    else :
        print "Graphviz OK!"

    # finish pre-check
    print "Pre-Check done!"
    return True

# ------------------------------------------------------------------ 
# Desc: CreateInstaller 
# ------------------------------------------------------------------ 

def CreateInstaller ():
    print "" 
    print "#########################" 
    print "installer"
    print "#########################" 
    print ""

    print "Creating installer"
    # KEEPME: no need { 
    # SafeCopyDirs ( os.path.join(exVim_path, "vim72"), os.path.join(dest_root_path,"rawdata/exvim/vim") )
    # os.makedirs( os.path.join(dest_root_path, "rawdata/exvim/vim/data/backup") )
    # os.makedirs( os.path.join(dest_root_path, "rawdata/exvim/vim/data/swap") )
    # } KEEPME end 

    # update vim-plugins
    print "Update vim-plugins..."

    vimplugins_path = os.path.join(dest_root_path,"rawdata/exvim/vim-plugins")  
    if os.path.isdir(vimplugins_path) == True :
        shutil.rmtree ( vimplugins_path )
    misc.SafeCopyDirs ( os.path.join(settings.exVim_path, "toolkit"), os.path.join(vimplugins_path,"toolkit") )
    misc.SafeCopyDirs ( os.path.join(settings.exVim_path, "vimfiles"), os.path.join(vimplugins_path,"vimfiles") )

    # update .vimrc
    print "Update .vimrc..."
    rawvim_path = os.path.join(dest_root_path,"rawdata/exvim/vim")
    if os.path.isdir(rawvim_path) == False :
        print "%s not found! can't copy .vimrc to it." % rawvim_path
    else :
        shutil.copy ( os.path.join(settings.exVim_path, ".vimrc"), rawvim_path )

    # update gvim.exe, ctags.exe 
    shutil.copy ( os.path.join(settings.exVim_path, "vim72/gvim_nopython.exe"), os.path.join(rawvim_path,"vim72/gvim.exe") )
    shutil.copy ( os.path.join(settings.exVim_path, "vim72/ctags.exe"), os.path.join(rawvim_path,"vim72/ctags.exe") )

    # update EnvVarUpdate.nsh
    print "Update EnvVarUpdate.nsh..."
    shutil.copy ( os.path.join(settings.exVim_path, "install/scripts/EnvVarUpdate.nsh"), dest_root_path )

    # update exvim.nsi
    print "Update exvim.nsi..."
    shutil.copy ( os.path.join(settings.exVim_path, "install/scripts/exvim.nsi"), dest_root_path )

# ------------------------------------------------------------------ 
# Desc: CompileInstaller 
# ------------------------------------------------------------------ 

def CompileInstaller ():
    # 
    installer_exe_path = os.path.join(dest_root_path, installer_exe_name + ".exe")
    installer_exe_ver_path = os.path.join(dest_version_path, installer_exe_name + "-" + settings.version + ".exe")

    # compile installer by exvim.nsi
    if os.path.isfile (installer_exe_ver_path) == True:
        print "%s already exists, if you want to re-generate one, remove it manually." % (installer_exe_name + "-" + settings.version)
        return

    cwd = os.getcwd()
    os.chdir( dest_root_path )
    os.system( 'makensis exvim.nsi')
    os.chdir( cwd )
    if os.path.isfile(installer_exe_path) == True :
        os.rename( installer_exe_path, installer_exe_ver_path) 
        print "Success: %s created!" % installer_exe_ver_path
    else :
        print "Failed: %s not found!" % installer_exe_name


#/////////////////////////////////////////////////////////////////////////////
# main
#/////////////////////////////////////////////////////////////////////////////

if __name__ == "__main__":
    if PreCheck() == True :
        CreateInstaller()
        CompileInstaller()

File: distribute/installer/gen_installer.lua
-- ======================================================================================
-- File         : gen_installer.lua
-- Author       : Wu Jie 
-- Last Change  : 01/04/2011 | 23:28:35 PM | Tuesday,January
-- Description  : 
-- ======================================================================================

--/////////////////////////////////////////////////////////////////////////////
-- requires
--/////////////////////////////////////////////////////////////////////////////

require "lfs"
require "ex.futil"
require "settings"

--/////////////////////////////////////////////////////////////////////////////
-- scripts
--/////////////////////////////////////////////////////////////////////////////

-- ------------------------------------------------------------------ 
-- Desc: precheck 
-- ------------------------------------------------------------------ 

local function precheck () 
    print "" 
    print "#########################" 
    print "Checking settings..."
    print "#########################" 
    print ""

    -- get version info
    print ("version = " .. settings.version)

    -- check if dest path exists
    print ("dest path = " .. settings.build_path)
    if lfs.attributes(settings.build_path) ~= nil then
        -- local ret,info = lfs.rmdir(settings.build_path)
        -- if ret == nil then
        --     print ("failed to remove directory " .. settings.build_path .. " the error message is: " .. info)
        --     return false
        -- end
        ex.futil.rmdirs(settings.build_path)
    end
    ret,info = lfs.mkdir(settings.build_path) -- create dest path
    if ret == nil then
        print ("failed to create dest path at " .. settings.build_path .. " the error message is: " .. info)
        return false
    end

    -- check exvim path
    print ("exvim path = " .. settings.exvim_path)
    if lfs.attributes(settings.exvim_path) == nil then
        print "can not found exvim"
        return false
    else
        print "exvim path OK!"
    end

    -- check exenv path
    print ("exenv path = " .. settings.exenv_path)
    if lfs.attributes(settings.exenv_path) == nil then
        print "can not found exenv"
        return false
    else
        print "exenv path OK!"
    end

    -- check graphviz path 
    print ("graphviz path = " .. settings.graphviz_path)
    if lfs.attributes(settings.graphviz_path) == nil then
        print "can not found graphviz"
        return false
    else
        print "graphviz path OK!"
    end

    -- finish pre-check
    print "done!"
    return true
end

-- ------------------------------------------------------------------ 
-- Desc: gen_installer 
-- ------------------------------------------------------------------ 

local function gen_installer () 
    print "" 
    print "#########################" 
    print "preparing install files..."
    print "#########################" 
    print ""

    print "copying vim73..."
    ex.futil.cpdir( settings.exvim_path.."/vim73", settings.build_path.."/rawdata/vim/vim73" )

    print "copying exvim..."
    ex.futil.cpdir( settings.exvim_path.."/toolkit", settings.build_path.."/rawdata/vim/exvim/toolkit" )
    ex.futil.cpdir( settings.exvim_path.."/vimfiles", settings.build_path.."/rawdata/vim/exvim/vimfiles" )
    ex.futil.cp( settings.exvim_path.."/.vimrc", settings.build_path.."/rawdata/vim/exvim" )

    print "copying resources..."
    ex.futil.cpdir( settings.exvim_path.."/distribute/installer/fonts", settings.build_path.."/rawdata/fonts" )
    ex.futil.cpdir( settings.exvim_path.."/distribute/installer/images", settings.build_path.."/rawdata/images" )

    print "copying gnu tools..."
    ex.futil.cpdir( settings.exenv_path.."/win32/bin", settings.build_path.."/rawdata/gnu/bin" )
    ex.futil.cpdir( settings.exenv_path.."/win32/share", settings.build_path.."/rawdata/gnu/share" )

    print "copying GraphViz..."
    ex.futil.cpdir( settings.graphviz_path, settings.build_path.."/rawdata/graphviz" )

    print "copying IrfanView..."
    ex.futil.cpdir( settings.irfanview_path, settings.build_path.."/rawdata/irfanview" )

    print "copying NSIS scripts..."
    ex.futil.cp( settings.exvim_path.."/distribute/installer/EnvVarUpdate.nsh", settings.build_path )
    ex.futil.cp( settings.exvim_path.."/distribute/installer/exvim.nsi", settings.build_path )
end

-- ------------------------------------------------------------------ 
-- Desc: 
-- ------------------------------------------------------------------ 

local function compile ()
    local dest = settings.target_path .. "/" .. settings.version
    ex.futil.mkdirs(dest)

    local cwd = lfs.currentdir ()
    lfs.chdir(settings.build_path)
    os.execute("makensis exvim.nsi")
    lfs.chdir(cwd)

    ex.futil.mv( settings.build_path .. "/exvim_installer.exe",
                 dest.."/exvim_installer-"..settings.version..".exe" )
end

-- ------------------------------------------------------------------ 
-- Desc: 
-- ------------------------------------------------------------------ 

precheck()
gen_installer()
compile()
